src/pointer_manager.h

```diff
@@ -0,0 +1,115 @@
+#pragma once
+
+#include <iostream>
+#include <map>
+#include <memory>
+#include <mutex>
+
+namespace grpc_labview
+{
+    //---------------------------------------------------------------------
+    // Definition
+    //---------------------------------------------------------------------
+
+    /// A pointer manager to track & manage the lifetime of pointers we expose to LabVIEW
+    /// This class serves two purposes:
+    /// 1. Ensure the validity of pointers -- This class knows which pointers are valid, so if
+    ///    LabVIEW passes us an invalid/stale pointer, we know it's invalid without ever trying to cast it.
+    /// 2. Gracefully clean up pointers only when they're no longer being used by utilizing RAII (std::shared_ptr).
+    ///    Client code never has to explicitly call `delete` on pointers. std::shared_ptr automatically calls
+    ///    delete when there are no remaining instances of std::shared_ptr for a given pointer. Usually, the
+    ///    call to UnregisterPointer() will remove the final instance of std::shared_ptr from the set, and that will
+    ///    cause `delete` to be called on that pointer. But if some other method in some other thread still has the std::shared_ptr
+    ///    from TryCastTo(), then `delete` won't be called until that other std::shared_ptr goes out of scope.
+    template <typename T>
+    class PointerManager
+    {
+    public:
+        PointerManager<T>();
+
+        /// Register a pointer. This creates a std::shared_ptr for that pointer and stuffs it into a std::set.
+        /// The pointer will be alive until that shared_ptr and all its copies (created via GetPointer) are out of scope.
+        /// Call UnregisterPointer to allow the pointer to be destroyed.
+        T* RegisterPointer(T* ptr);
```

Comment:
Since we are using shared_ptr semantics for lifetime management, should we take in a shared pointer for registration?

I can go either way on this.  I realize that given how we are using things with CLN we are passing pointers around and Unregister makes more sense to take a pointer.

Reply:
eh, I am just going to submit as it. We can think about it some more and do a follow up if we think it is warranted.

Reply:
I think I like it just taking a raw pointer, but I could go either way as well.

I also thought about not taking a pointer at all and just telling the PointerManager to create the object.

i.e.
`grpc_labview::gPointerManager.CreateAndRegisterPointer<grpc_labview::ClientCall>(timeoutMs)`
instead of
`*callId = grpc_labview::gPointerManager.RegisterPointer(new grpc_labview::ClientCall(timeoutMs));`).

Maybe even make all the constructors for derived gRPCid classes protected so that pointers can only be created by that factory method. But that seemed like overkill.

Like you said, since we're passing naked pointers to and from the CLNs, I vote we just expose the naked pointers.